esp-hal/src/dma/buffers.rs

```diff
@@ -5,25 +5,65 @@ use crate::soc::is_slice_in_dram;
 #[cfg(esp32s3)]
 use crate::soc::is_slice_in_psram;
 
+/// Burst transfer configuration.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum BurstTransfer {
+    /// Burst mode is disabled.
+    Disabled,
+
+    /// Burst mode is enabled.
+    Enabled,
+}
+
+impl BurstTransfer {
+    pub(super) fn is_burst_enabled(self) -> bool {
+        !matches!(self, Self::Disabled)
+    }
+}
+
+/// Maximum length of a burst transfer.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum TransferDirection {
+    /// DMA transfer from peripheral or external memory to memory.
+    In,
+    /// DMA transfer from memory to peripheral or external memory.
+    Out,
+}
+
 /// Holds all the information needed to configure a DMA channel for a transfer.
 pub struct Preparation {
     /// The descriptor the DMA will start from.
     pub start: *mut DmaDescriptor,
 
-    /// Block size for PSRAM transfers
-    #[cfg_attr(not(esp32s3), allow(dead_code))]
-    pub block_size: Option<DmaBufBlkSize>,
+    /// The direction of the DMA transfer.
+    pub direction: TransferDirection,
```

Comment:
Call this self-defense, I'll screw something up for sure around this if we don't have a way to check.

Reply:
So once you're done refactoring, we can remove it?

(I still think it's a strange thing to have to specify as a DMA buffer trait implementer)

Reply:
I don't intend to, this is a pretty cheap way to declare that indeed, a buffer is being sent to the intended channel. As preparation isn't unsafe, and is publicly available, I want at least a bit of safety in case someone ends up using it barefeet.

Reply:
That, or we can have InPreparation and OutPreparation, or a direction type param on Preparation.

Reply:
> That, or we can have InPreparation and OutPreparation, or a direction type param on Preparation.

To be honest, I'd like that much better than this enum. (Might even work out nicely for the P4's DMA weirdness)

Reply:
> As preparation isn't unsafe

It is. The entire trait being unsafe means preparation is unsafe. There's no other way in Rust (that I know of) to make a trait method unsafe for the implementer (instead of the caller).

Reply:
Remind me when we get closer to the later PRs, please.

Reply:
> It is.

The trait is, the structure isn't. The structure could be turned into a non-exhaustive struct with an unsafe constructor, I guess, for third parties.

Reply:
Let's explore refactoring this out in later